ldap/server.go

```diff
@@ -80,15 +82,66 @@ type Server struct {
 	RootDSE map[string][]string
 
 	tlsConfig *tls.Config
+	wg        *waitGroup
+	mu        sync.Mutex
+	listeners []net.Listener
+	isStopped bool
+	factory   listenerFactory
+}
+
+// waitGroup - kind of a std WaitGroup
+// we can't use std WaitGroup here to keep original Server API
+// std WaitGroup doesn't allow to call Wait before Add
+// but technically we can call Shutdown before Serve
```

Comment:
> but technically we can call Shutdown before Serve

but why do we need to call it in that way?

Reply:
Just because it's possible:-)